lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HDFSCommandRunner.cs

```diff
@@ -0,0 +1,172 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using Org.Apache.REEF.IO.FileSystem.Hadoop.Parameters;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.IO.FileSystem.Hadoop
+{
+    /// <summary>
+    /// Helper class to execute commands
+    /// </summary>
+    internal sealed class HdfsCommandRunner
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(HdfsCommandRunner));
+
+        /// <summary>
+        /// Path to hdfs.cmd
+        /// </summary>
+        private readonly string _hdfsCommandPath;
+
+        /// <summary>
+        /// The number of retries on HDFS commands.
+        /// </summary>
+        private readonly int _numberOfRetries;
+
+        /// <summary>
+        /// The timeout on each of the retries.
+        /// </summary>
+        private readonly int _timeOutInMilliSeconds;
+
+        [Inject]
+        private HdfsCommandRunner([Parameter(typeof(NumberOfRetries))] int numberOfRetries,
+            [Parameter(typeof(CommandTimeOut))] int timeOutInMilliSeconds,
+            [Parameter(typeof(HadoopHome))] string hadoopHome)
+        {
+            _numberOfRetries = numberOfRetries;
+            _timeOutInMilliSeconds = timeOutInMilliSeconds;
+
+            if (hadoopHome == HadoopHome.DefaultValue)
+            {
+                var hadoopHomeFromEnv = Environment.GetEnvironmentVariable("HADOOP_HOME");
+                if (null == hadoopHomeFromEnv)
+                {
+                    throw new Exception("HADOOP_HOME not set and no path to the hadoop installation provided.");
+                }
+                _hdfsCommandPath = Path.Combine(hadoopHomeFromEnv, "bin", "hdfs.cmd");
+            }
+            else
+            {
+                _hdfsCommandPath = Path.Combine(hadoopHome, "bin", "hdfs.cmd");
+            }
+
+
+            if (!File.Exists(_hdfsCommandPath))
+            {
+                throw new Exception("HDFS command does not exist: " + _hdfsCommandPath);
+            }
+        }
+
+        internal CommandResult Run(string hdfsCommandLineArguments)
+        {
+            var outList = new List<string>();
+            var errList = new List<string>();
+            var tries = _numberOfRetries;
+            var origTries = _numberOfRetries;
+            Process process;
+            do
+            {
+                outList.Clear();
+                Logger.Log(Level.Info, "Trial {0} Timeout in {1} secs Executing: {2}", origTries - tries,
+                    _timeOutInMilliSeconds/1000,
+                    _hdfsCommandPath + " " + hdfsCommandLineArguments);
+                var startInfo = new ProcessStartInfo
```

Comment:
Would it be better to use new Process and register the callbacks first instead starting with ProcessStartInfo and then registering the callbacks?